core/src/main/groovy/com/craigburke/document/core/Image.groovy

```diff
@@ -9,9 +11,45 @@ class Image extends BaseNode {
     ImageType type = ImageType.JPG
     Integer width
     Integer height
-    byte[] data
-    
-    void setType(String value) { 
-        type = Enum.valueOf(ImageType, value.toUpperCase()) 
+    private URL imageUrl
+    private byte[] imageData
+
+    void setType(String value) {
+        type = Enum.valueOf(ImageType, value.toUpperCase())
+    }
+
+    URL getUrl() {
+        imageUrl
+    }
+
+    void setUrl(String value) {
+        setUrl(value != null ? URI.create(value).toURL() : null)
+    }
+
+    void setUrl(URL value) {
+        imageUrl = value
+    }
+
+    byte[] getData() {
+        imageUrl?.bytes ?: imageData
+    }
+
+    void setData(byte[] data) {
+        imageData = Arrays.copyOf(data, data.length)
+    }
+
+    def withInputStream(Closure work) {
+        if(imageUrl != null) {
+            return imageUrl.withInputStream(work)
+        }
+        work.call(new ByteArrayInputStream(imageData))
+    }
+
+    String getHash() {
```

Comment:
@craigburke What do you think about the Arrays.copyOf(...) ? Am I being to paranoid about mutability? I wanted to write safe code but on the other hand I don't want to waste memory. I'm sure someone will create documents with lots of images.

Another related question is about if the first access to url should store the bytes in imageData. That way you only get one network request even when you use the image multiple times in a document. My current impl request the image bytes from the url every time you call getData().